code-studio/scripts/build-commands.js

```diff
@@ -154,57 +163,75 @@ exports.bundle = function (config) {
- * @param {string[]} commands - array of shell commands to be executed in sequence.
+ * Executes a single shell command in a child process that inherits the
+ * environment and stdout/stderr from this process.
+ *
+ * @param {!string} command
+ * @returns {Promise.<string>} which resolves to the child process' stdout if
+ * the command exits with code zero, or rejects if the child process fails.
  */
-exports.execute = function (commands) {
-  commands.forEach(function (command) {
-    try {
-      console.log(command);
-
-      // For documentation on synchronous execution of shell commands from node scripts, see:
-      // https://nodejs.org/docs/latest/api/child_process.html#child_process_synchronous_process_creation
-      var result = child_process.execSync(command, {
-        env: _.extend({}, process.env, {
-          PATH: './node_modules/.bin:' + process.env.PATH
-        }),
-        stdio: 'inherit'
-      });
-    } catch (e) {
-      exports.exitWithError(e);
-    }
+exports.executeShellCommand = function (command) {
+  return new Promise(function (resolve, reject) {
+    console.log(command);
+    child_process.exec(command, getChildProcessOptions(), function (err, stdout) {
+      if (err) {
+        reject(transformExecError(err));
+      } else {
+        resolve(stdout);
+      }
+    });
   });
 };
 
 /**
- * End the node process with the given error message and code.
- * Ensures an exit code of at least one, in case the error doesn't have a code.
- * @param {Error} err
+ * Given an array of shell commands, executes those commands in child processes
+ * in parallel.
+ * @param {!string[]} commands - list of shell commands
+ * @returns {Promise} which resolves when all commands are completed, or rejects
+ *          as soon as any command fails.
  */
-exports.exitWithError = function (err) {
-  console.log("\nError: " + err.message);
-  warnIfWrongNodeVersion();
-  process.exit(err.status || 1);
+exports.executeShellCommandsInParallel = function (commands) {
+  return Promise.all(commands.map(exports.executeShellCommand));
+};
+
+/**
+ * Given an array of functions that return a promise, invokes the functions
+ * in sequence, waiting to invoke each function until the promise returned by
+ * the function before it in the list has resolved.  If any promise in the
+ * sequence rejects, the rest of the sequence is skipped.
+ *
+ * This is an alternative to manually chaining these functions.  Instead of:
+ *     fn1().then(fn2).then(fn3).then(fn4).then(done);
+ * You can call:
+ *     promiseSequence([fn1, fn2, fn3, fn4]).then(done);
+ *
+ * @param {function[]} functionSequence - functions that return Promises, to be
+ *        called in sequence.
+ * @returns {Promise} which resolves when the Promise returned by the last
+ *          function resolves, or rejects as soon as any Promise in the
+ *          sequence rejects.
+ */
+exports.promiseSequence = function (functionSequence) {
```

Comment:
Oops!  I was using it in my `executeShellCommandSequence` method, but _that_ wasn't being used anywhere yet (turns out the places I build up more than one or two commands I want to run them in parallel) so... no, not used yet.  I guess I should tear this out?